### PR TITLE
User Management with NgRx integration and Auth Guards

### PR DESCRIPTION
**Pull request by tanadendere**
This pull request introduces the user management system integration with NgRx store. It also introduces authentication guards.

Changes Made:

- Integration between the user management and NgRx.
- Added auth guards.
- Successful happy path scenarios for both login and registration processes (still to do validation and other error handling in another PR).

### CODE REVIEW

File: src/app/app.component.html [hunk 1]
```diff
@@ -1,13 +1 @@
...
-  <a routerLink="/register"> Register</a>
-</div>
-} @if (authService.currentUserSignal()){
-<div>
-  {{ authService.currentUserSignal()?.name }}
-  <span (click)="logout()">Logout</span>
-</div>
-}
-
 <router-outlet></router-outlet>
```
Comment (Musa-Mabasa) [inline src/app/app.component.html]: A suggestion, You can make this self enclosing: `<router-outlet/>`

File: src/app/components/home/home.component.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+@Component({
+  selector: 'app-home',
+  standalone: true,
+  imports: [CommonModule, RouterOutlet, RouterLink],
+  templateUrl: './home.component.html',
+  styleUrl: './home.component.scss',
+})
+export class HomeComponent {
+  store: Store<AppState> = inject(Store);
+  user$: Observable<IUser | undefined> = this.store.select(selectUser);
```
Comment (Musa-Mabasa) [inline src/app/components/home/home.component.ts]: I think the type is inferred here. So there variables do not need to be typed

File: src/app/userManagement/components/register/register.component.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    this.store.dispatch(
+      registerUser({
+        email: rawForm.email,
+        name: rawForm.name,
+        surname: rawForm.surname,
+        password: rawForm.password,
+      })
+    );
+
+    this.user$.subscribe((user) => {
```
Comment (Musa-Mabasa) [inline src/app/userManagement/components/register/register.component.ts]: Remember to destroy subscriptions

File: src/app/userManagement/services/crud.service.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+          user.email
+        ),
+        {
+          email: user.email,
+          name: user.name,
+          password: user.password,
+          surname: user.surname,
+          userId: user.userId,
+        }
+      ).catch((error) => console.error('Error adding user: ', error))
```
Comment (Musa-Mabasa) [inline src/app/userManagement/services/crud.service.ts:29]: Nice!! Error handling 👍

File: src/app/userManagement/store/effects.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    this.actions$.pipe(
+      ofType(loginUser.type),
+      switchMap((action) =>
+        this.authService.login(action.email, action.password).pipe(
+          map(() => {
+            return getUserFromFirestore({ email: action.email });
+          }),
+          retry(1),
+          catchError((err) => {
+            alert(`This user does not exit` + err.toString());
```
Comment (Musa-Mabasa) [inline src/app/userManagement/store/effects.ts]: A typo here for exist I think.

File: src/app/app.routes.ts [hunk 6]
```diff
@@ -1,14 +1,39 @@
...
+    path: '',
+    component: LoginComponent,
+    pathMatch: 'full',
+    canActivate: [AuthGuard],
+  },
+  {
+    path: '**',
+    component: LoginComponent,
+    pathMatch: 'full',
+    canActivate: [AuthGuard],
```
Comment (yanevdw) [inline src/app/app.routes.ts]: Perhaps consider `redirectTo` or a`NotFoundPage`. 

Also, with regards to the `AuthGuard`, why is it required by your Login page or the wildcard route? 

Normally, `canActivate` is used for your Home page for example, but would this implementation not also prevent access to your Login page?

File: src/app/userManagement/components/register/register.component.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  user$: Observable<IUser | undefined> = this.store.select(selectUser);
+
+  hasUnsavedChanges = true;
+
+  form = this.fb.nonNullable.group({
+    email: ['', Validators.required],
+    name: ['', Validators.required],
+    surname: ['', Validators.required],
+    password: ['', Validators.required],
+    confirmPassword: ['', Validators.required],
```
Comment (yanevdw) [inline src/app/userManagement/components/register/register.component.ts:31]: You could add patterns here to match the password with regex.

**Review by yanevdw — COMMENTED** [review]
Fantastic job with the NgRx integration! ✨ Just the general suggestions and the comment about protected routes.

**Review by yanevdw — APPROVED** [review]

File: src/app/userManagement/store/effects.ts (hunk 5 above)
Comment (tanadendere) [reply src/app/userManagement/store/effects.ts]: Yes oh my gosh thank you 😂

File: src/app/app.routes.ts (hunk 6 above)
Comment (tanadendere) [reply src/app/app.routes.ts]: Okay cool, i've changed to redirectTo and will do the NotFoundPage in the error handling PR.

Oh my `canActivate` function always returns true (for now). But when they have the app installed on the phone i would like for the user to be able to just go straight to the home page... thats why I implemented it that way.

File: src/app/userManagement/components/register/register.component.ts (hunk 7 above)
Comment (tanadendere) [reply src/app/userManagement/components/register/register.component.ts:31]: Yea i will do so in the Error handing PR, I am just focusing on a happy path for now

**Review by Musa-Mabasa — APPROVED** [review]
Some minor suggestions here and there. Everything looks good 👍

**Pull request merged**